### PR TITLE
Fix failures with latest responses library (0.12.1)

### LINKED ISSUE #3460: Does not work with latest `responses`
Moto has a hack where it sets the default responses mock's `_find_match` method to a custom one. The latest changes in `responses` have changed the return type of this method.

https://github.com/spulec/moto/blob/47dbad291e717059a93f5cf795a9be08cfc872cc/moto/core/models.py#L248-L267

See the [`responses`](https://github.com/getsentry/responses) change here https://github.com/getsentry/responses/commit/cf3c5a685ae95bee1a714fd36cfb77d448a4b5ec

### PR DESCRIPTION
**Pull request by garribas**
Fixes #3460 

Before:
```
$ pip freeze | grep responses
responses==0.12.1

$make test
...
Ran 3601 tests in 705.949s

FAILED (SKIP=1, errors=28, failures=6)
```

After:
```
$ pip freeze | grep responses
responses==0.12.1

$make test
...
Ran 3601 tests in 591.731s

OK (SKIP=1)
```

### CODE REVIEW

**Comment by davidszotten** [comment]
should we consider trying to detect version or return type rather than forcing the latest version of the library?

**Comment by ypankovych** [comment]
Is this gonna bump the major version?

**Comment by garribas** [comment]
@davidszotten imho implementing a monkey patch that deals with different versions of a third party library is bad long term from a maintainability perspective.

@ypankovych I'm not familiar with moto's release process and how versions are bumped, happy to bump version if you point me how to do it.

Current and past versions of moto are currently broken if doing a fresh install. If this approach isn't enticing we can instead change setup.py to not install responses versions post 0.12.0.

**Review by bblommers — CHANGES_REQUESTED** [review]
Thanks for the PR @garribas! I agree with @davidszotten on this - I'd prefer a solution where we detect the return type.

> imho implementing a monkey patch that deals with different versions of a third party library is bad long term from a maintainability perspective.

Agreed, but it's the lesser of two evils IMO. There are so many projects using moto, there will always be some who can't upgrade a specific dependency. Having some techdebt is preferable to locking people out from using Moto altogether.

**Comment by garribas** [comment]
@bblommers I've updated the PR to detect version (detecting return type looked too fragile), relies on pkg_version (setuptools dependency) and distutils (built-in) to detect responses lib version.

**Comment by davidszotten** [comment]
1) i wonder if we just want to different versions of the whole function, and then detect version before monkeypatching instead of having to check (twice) at runtime. i think this would be more readable

2) consider using importlib.metadata (falling back to the backport importlib-metadata) which is way faster than pkg_resources

File: moto/core/models.py [hunk 1]
```diff
@@ -252,12 +255,22 @@ def _url_matches(self, url, other, match_querystring=False):
 
 
 def _find_first_match(self, request):
+    match_failed_reasons = []
     for i, match in enumerate(self._matches):
-        if match.matches(request):
-            return match
-
-    return None
+        if LooseVersion(RESPONSES_VERSION) < LooseVersion('0.12.1'):
```
Comment (bblommers) [inline moto/core/models.py:260]: For readability, would you mind rewriting this to only have the single if-statement? 
Whether it's a separate function or everything in the same function is the same to me.

Usage of 'pkg_resources' is fine IMO.

You might have to run 'black' as well afterwards, as Travis is currently complaining about it.

**Review by bblommers — CHANGES_REQUESTED** [review]

**Review by bblommers — APPROVED** [review]
LGTM. Thank you so much @garribas!

**Comment by mludvig** [comment]
Confirming that this change fixes our problem with `responses==0.12.1`, thanks!

Any chance this can be released to pypi sometime soon?

**Comment by bblommers** [comment]
A dev-release is already available on PyPi, @mludvig - it is part of moto >= 1.3.16.dev122

**Comment by flying-sheep** [comment]
I don’t get it. 1.3.16 was released in september and `1.3.16.dev122 < 1.3.16`.

Does moto use the wrong version tags?

**Pull request merged**